File: 2D_sym_python/Animal.py
from Organism import Organism
from Defines import*

class Animal(Organism):

    def __init__(self):
        super().__init__()
        self.initiative = 0
        self.strength = 0
        self.name = ''
        self.isAlive = True
    
    def action(self, world):
        world.moveOrganism(self, ONE_STEP)
    
    def collision(self,world,organismToFight):
        if(self.isAttacker==True):
            self.defaultAttack(world, organismToFight)
            return True
        return False
    
    def defaultAttack(self, world, organismToFight):
        if organismToFight.GetOrganismLifespan()==True:

            if self.name == organismToFight.GetName():
                if world.addConcreteOrganism(self, organismToFight, NEW_ORGANISM):
                    world.AddCommentatorVector(reportType.SPAWN, self, None, 0)
                self.SetOrganismWantedPosition(0, 0)
            
            elif not organismToFight.collision(world, self):
                if self.strength == organismToFight.GetStrength() or self.strength > organismToFight.GetStrength():
                    organismToFight.SetOrganismLifespan(False)

                    world.SetOrganismVectorData(None, self.position['y'], self.position['x'])
                    self.SetOrganismPosition(self.wantedPos['x'],self.wantedPos['y'])
                    self.SetOrganismWantedPosition(0, 0)

                    world.SetOrganismVectorData(self, self.position['y'], self.position['x'])
                    world.AddCommentatorVector(reportType.ANIMAL_OR_PLANT, self, organismToFight, 0)
                else:
                    self.isAlive = False
                    self.SetOrganismWantedPosition(0, 0)
                    world.SetOrganismVectorData(None, self.position['y'], self.position['x'])
                    world.AddCommentatorVector(reportType.ANIMAL_OR_PLANT, organismToFight, self, 0)
        
        else:
            world.SetOrganismVectorData(None,self.position['y'],self.position['x'])
            self.SetOrganismPosition(self.wantedPos['x'],self.wantedPos['y'])
            self.SetOrganismWantedPosition(0,0)
            world.SetOrganismVectorData(self, self.position['y'],self.position['x'])

            world.addCommentatorVector(reportType.MOVED, self,None,0)
        return True
        

File: 2DVirtualWorldSimulatorCpp/2DVirtualWorldSimulator/Animal.cpp
#include <iostream>
#include <cstdlib>
#include "Organism.h"
#include "Animal.h"
#include "World.h"
#include "Defines.h"

Animal::Animal() {
	this->name = NULL;
	this->strength = 0;
	this->initiative = 0;
	this->position.x = 0;
	this->position.y = 0;
	this->wantedPos.x = NULL;
	this->wantedPos.y = NULL;
	this->isAlive = true;
}

void Animal::action(World* world) {
	world->moveOrganism(this, ONE_STEP);
}

bool Animal::collision(World* world,Organism* organismToFight) {
	if (this->isAttacker == true) {
		defaultAttack(world, organismToFight);
		return true;
	}
	return false;
}

bool Animal::defaultAttack(World* world, Organism* organismToFight) {
	if (organismToFight->GetOrganismLifespan() == true) {
		
		if (this->name == organismToFight->GetName()) {
			if (world->addConcreteOrganism(this, organismToFight, NEW_ORGANISM)) {
				world->addCommentatorVector(SPAWN, this, nullptr, NULL);
			}
			this->SetOrganismWantedPosition(NULL, NULL);
		}
		else if (!organismToFight->collision(world, this)) {
			if ((this->strength == organismToFight->GetStrength()) || (this->strength > organismToFight->GetStrength())) {
				organismToFight->SetOrganismLifespan(false);

				world->setOrganismVectorData(nullptr, position.y, position.x);
				this->SetOrganismPosition(wantedPos.x, wantedPos.y);
				this->SetOrganismWantedPosition(NULL, NULL);

				world->setOrganismVectorData(this, position.y, position.x);
				world->addCommentatorVector(ANIMAL_OR_PLANT, this, organismToFight, NULL);
			}
			else {
				this->isAlive = false;
				this->SetOrganismWantedPosition(NULL, NULL);
				world->setOrganismVectorData(nullptr, position.y, position.x);
				world->addCommentatorVector(ANIMAL_OR_PLANT, organismToFight, this, NULL);
			}
		}
	}
	else {
		world->setOrganismVectorData(nullptr, position.y, position.x);
		this->SetOrganismPosition(wantedPos.x, wantedPos.y);
		this->SetOrganismWantedPosition(NULL, NULL);
		world->setOrganismVectorData(this, position.y, position.x);

		world->addCommentatorVector(MOVED, this, nullptr, NULL);
	}
	return true;
}

int Animal::GetFeatures()const {
	static int numFeature = 0;
	int out = NULL;
	if (numFeature == 0) out = this->name, numFeature++;
	else if (numFeature == 1) out = this->strength, numFeature++;
	else if (numFeature == 2) out = this->position.x, numFeature++;
	else if (numFeature == 3) out = this->position.y, numFeature++;
	else if (numFeature == 4) out = this->wantedPos.x, numFeature++;
	else if (numFeature == 5) out = this->wantedPos.y, numFeature++;
	else if (numFeature == 6) {
		numFeature = 0;
		if (this->isAlive) out = 1;
		else out = 0;
	}
	return out;
}

void Animal::SetFeatures(int feature_, World* world) {
	static int numFeature = 0;
	static pos positionTemp;

	if (numFeature == 0) this->SetOrganismStrength(feature_) , numFeature++;
	else if(numFeature == 1) positionTemp.x = feature_, numFeature++;
	else if (numFeature == 2) positionTemp.y = feature_, numFeature++, this->SetOrganismPosition(positionTemp.x, positionTemp.y);
	else if (numFeature == 3) positionTemp.x = feature_, numFeature++;
	else if (numFeature == 4) positionTemp.y = feature_, numFeature++, this->SetOrganismWantedPosition(positionTemp.x, positionTemp.y);
	else if (numFeature == 5) {
		numFeature = 0;
		if (feature_ == 1) this->SetOrganismLifespan(true);
		else this->SetOrganismLifespan(false);
	}
}

Animal::~Animal() {

}